src/workerd/api/sockets.c++

```diff
@@ -511,4 +511,93 @@ jsg::Ref<Socket> SocketsModule::connect(
+      kj::StringPtr url,
+      const kj::HttpHeaders& headers,
+      kj::AsyncInputStream& requestBody,
+      kj::HttpService::Response& response) override {
+    // Parse the URL to extract the path
+    auto parsedUrl = KJ_REQUIRE_NONNULL(
+        kj::Url::tryParse(url, kj::Url::Context::HTTP_PROXY_REQUEST), "invalid url", url);
+
+    // Create a new HTTP client using our stream
+    auto httpClient = kj::newHttpClient(headerTable, *stream);
+
+    // Create a new HTTP service from the client
+    auto service = kj::newHttpService(*httpClient);
+
+    // Prepare headers for the request
+    auto newHeaders = headers.cloneShallow();
+    newHeaders.set(kj::HttpHeaderId::HOST, parsedUrl.host);
+
+    // Get just the path portion for the request
+    auto pathUrl = parsedUrl.toString(kj::Url::Context::HTTP_REQUEST);
+
+    // Forward the request to the service
+    return service->request(method, pathUrl, newHeaders, requestBody, response)
+        .attach(kj::mv(service), kj::mv(httpClient));
+  }
+
+  kj::Promise<void> connect(kj::StringPtr host,
+      const kj::HttpHeaders& headers,
+      kj::AsyncIoStream& connection,
+      ConnectResponse& response,
+      kj::HttpConnectSettings settings) override {
+    KJ_UNIMPLEMENTED("connect() not supported on StreamWorkerInterface");
+  }
+
+  kj::Promise<void> prewarm(kj::StringPtr url) override {
+    KJ_UNIMPLEMENTED("prewarm() not supported on StreamWorkerInterface");
+  }
+
+  kj::Promise<ScheduledResult> runScheduled(kj::Date scheduledTime, kj::StringPtr cron) override {
+    KJ_UNIMPLEMENTED("runScheduled() not supported on StreamWorkerInterface");
+  }
+
+  kj::Promise<AlarmResult> runAlarm(kj::Date scheduledTime, uint32_t retryCount) override {
+    KJ_UNIMPLEMENTED("runAlarm() not supported on StreamWorkerInterface");
+  }
+
+  kj::Promise<CustomEvent::Result> customEvent(kj::Own<CustomEvent> event) override {
+    return event->notSupported();
+  }
+
+ private:
+  kj::Own<kj::AsyncIoStream> stream;
+  const kj::HttpHeaderTable& headerTable;
+};
+
+// Implementation of the custom factory for creating WorkerInterface instances from a socket
+class SocketOutgoingFactory final: public Fetcher::CrossContextOutgoingFactory {
+ public:
+  SocketOutgoingFactory(jsg::Ref<Socket> socket): socket(kj::mv(socket)) {}
+
+  kj::Own<WorkerInterface> newSingleUseClient(
+      IoContext& context, kj::Maybe<kj::String> cfStr) override {
+    // Create a WorkerInterface that wraps the stream
+    // TODO(now): Is context.getHeaderTable() right here?
+    return kj::heap<StreamWorkerInterface>(socket->getConnectionStream(), context.getHeaderTable())
+        .attach(socket.addRef());
+  }
+
+ private:
+  jsg::Ref<Socket> socket;
+};
+
+jsg::Ref<Fetcher> SocketsModule::createHttpClient(jsg::Lock& js, jsg::Ref<Socket> socket) {
+  // Create our custom factory that will create client instances from this socket
+  auto outgoingFactory = kj::heap<SocketOutgoingFactory>(socket.addRef());
```

Comment:
This should check to make sure the `socket`'s `readable` or `writable` are not locked, and possibly not disturbed. It especially needs to check to ensure the socket is not closed, and probably needs to wait for the opened promise to be resolved.

Locked indicates that there is `Reader` or `Writer` associated, in which case wrapping the socket should fail. Disturbed indicates that the socket has been written to or read from, indicating that it may not be in the correct state to be wrapped correctly.

Reply:
We should probably also arrange it such that the returned `Fetcher` completely takes ownership over the `Socket` such that the `readable` and `writable` streams get closed/detached or locked so that nothing on the JS side mistakenly tries to access them while the socket it bound... e.g.

```js
const socket = connect('...');
const fetcher = getHttpClient(socket);
socket.readable.getReader();  // should fail!
```

Reply:
As an additional bit of fun, since `fetch(...)` is cancelable using an AbortSignal, the tests need to verify that canceling a fetch via the returned Fetcher works as expected here. 

Also, is the `Socket` expected to support multiple http requests or should it be closed when the one http request is finished? Would it be expected to support pipelining (I'd expect no). This would need to ensure that we don't end up with multiple concurrent http requests going over the socket at a time... e.g. 

```js
const socket = connect('...');
const fetcher = getHttpClient(socket);
await Promise.all([ fetcher.fetch('...'), fetcher.fetch('...') ]);
```

Reply:
A few comments here.
Regarding the state of socket when receiving control and after. I left it a bit vague because this is intended to be an internal API only used by one worker really, I don't mind amending the implementation to make it a bit more safe in this regard though.

Socket is expected to support multiple http requests and there's even a test case for it.